documentation/tuning_guide.markdown

```diff
@@ -0,0 +1,156 @@
+environments to a single pool, to ensure that there is no pollution from one
+Puppet environment to the next.)
+
+This brings us to the two most important settings that you can use to tune your
+Puppet Server.
+
+### Number of JRubies
+
+The most important setting that you can use to improve the throughput of your
+Puppet Server installation is the [`max-active-instances`](./configuration.html#puppetserver_conf)
+setting.  The value of this setting is used by Puppet Server to determine how
+many JRuby instances to create when the server starts up.
+
+From a practical perspective, this setting basically controls how many Puppet
+agent runs Puppet Server can handle concurrently.  The minimum value you can
+get away with here is `1`, and if your installation is small enough such that
+you're unlikely to ever have more than one Puppet agent checking in with the
+server at exactly the same time, this will be totally sufficient.
+
+However, if you specify a value of `1` for this setting, and then you have two
+Puppet agent runs hitting the server at the same time, what will happen is that
+the requests being made by the second agent will be effectively blocked until
+the server has finished handling all of the requests from the first agent.
+(In other words, one of Puppet Server's threads will have "borrowed" the single
+JRuby instance from the pool to use to handle the requests from the first agent,
+and only when those requests are completed will it return the JRuby instance
+to the pool.  At that point another thread can "borrow" the JRuby instance to
+use to handle the requests from the second agent.)
+
+Assuming you have more than one CPU core in your machine, this situation means
+that you won't be getting the maximum possible throughput from your Puppet Server
+installation.  Increasing the value from `1` to `2` would mean that Puppet Server
+could now use a second CPU core to handle the requests from a second Puppet agent
+simultaneously.
+
+It follows, then, that the maximum sensible value to use for this setting will
+be roughly the number of CPU cores you have in your server.  Setting the value
+to something much higher than that won't improve performance, because even if there
+are extra JRuby instances available in the pool to do work, they won't be able
+to actually do any work if all of the CPU cores are already busy using JRuby
+instances to handle incoming agent requests.
+
+(There are exceptions to this rule; if you have report processors that make
+a network connection as part of the processing of a report, and if there is a chance
+that the network operation is slow and will block on I/O for some period of time,
+then it may make sense to have more JRuby instances than the number of cores,
+because the JVM will be smart enough to suspend the thread that is handling those
+kinds of requests and use the CPUs for other work, assuming there are still JRuby
+instances available in the pool.  In a case like this you might want to set
```

Comment:
Not sure if this would be obvious to the reader, but if a thread holding a JRuby instance were suspended, the JRuby instance would not implicitly be returned to the pool while suspended and reacquired when the thread is resumed later.  The JRuby instance would only be freed up around the time that the response from the request is being delivered by the master.